File: src/main/java/codereview/simpleorder/order/application/OrderService.java
package codereview.simpleorder.order.application;

import codereview.simpleorder.order.domain.Order;
import codereview.simpleorder.order.domain.OrderItem;
import codereview.simpleorder.order.dto.web.CreateOrderRequest;
import codereview.simpleorder.order.dto.web.ItemResponse;
import codereview.simpleorder.order.dto.web.OrderLineRequest;
import codereview.simpleorder.order.repository.OrderRepository;
import codereview.simpleorder.order.rest.ItemFeignClient;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static java.util.stream.Collectors.toList;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class OrderService {

    private final OrderRepository orderRepository;
    private final ItemFeignClient itemHttpClient;

    @Transactional
    public Long order(CreateOrderRequest orderRequest) {

        List<OrderLineRequest> orderLineRequests = orderRequest.getOrderLineRequests();
        List<Long> itemIds = extractItemIds(orderLineRequests);

        List<ItemResponse> itemResponses = itemHttpClient.requestItems(itemIds).getItems();

        List<OrderItem> orderItems = createOrderLines(itemResponses, orderLineRequests);
        Order order = Order.createOrder(orderItems);
        Order savedOrder = orderRepository.save(order);

        return savedOrder.getId();
    }

    private static List<Long> extractItemIds(List<OrderLineRequest> orderLineRequests) {

        return orderLineRequests.stream()
                .map(OrderLineRequest::getItemId)
                .collect(toList());
    }

    private List<OrderItem> createOrderLines(List<ItemResponse> itemResponses, List<OrderLineRequest> orderLineRequests) {

        return orderLineRequests.stream()
                .map(request -> createOrderLine(itemResponses, request))
                .collect(toList());
    }

    private OrderItem createOrderLine(List<ItemResponse> itemResponses, OrderLineRequest orderLineRequest) {

        ItemResponse itemResponse = findItemDtoFromOrderLineRequest(itemResponses, orderLineRequest);

        return new OrderItem(
                itemResponse.getId(),
                itemResponse.getName(),
                itemResponse.getSize(),
                itemResponse.getPrice(),
                orderLineRequest.getQuantity());
    }

    private static ItemResponse findItemDtoFromOrderLineRequest(List<ItemResponse> itemResponses, OrderLineRequest orderLineRequest) {

        return itemResponses.stream()
                .filter(it -> it.getId().equals(orderLineRequest.getItemId()))
                .findAny()
                .orElseThrow(() -> new IllegalArgumentException("주문 항목 요청에 해당하는 상품이 없습니다."));
    }
}

File: order/src/main/kotlin/msa/with/ddd/order/application/OrderService.kt
package msa.with.ddd.order.application

import msa.with.ddd.order.domain.Order
import msa.with.ddd.order.domain.OrderItem
import msa.with.ddd.order.dto.web.CreateOrderRequest
import msa.with.ddd.order.dto.web.ItemResponse
import msa.with.ddd.order.dto.web.OrderLineRequest
import msa.with.ddd.order.repository.OrderRepository
import msa.with.ddd.order.rest.ItemFeignClient
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
class OrderService(
    private val orderRepository: OrderRepository,
    private val itemHttpClient: ItemFeignClient
) {

    @Transactional
    fun order(orderRequest: CreateOrderRequest): Long {
        val orderLineRequests = orderRequest.orderLineRequests
        val itemIds = extractItemIds(orderLineRequests)
        val itemResponses = itemHttpClient.requestItems(itemIds).items
        val orderItems = createOrderLines(itemResponses, orderLineRequests)
        val order = Order.createOrder(orderItems)
        val savedOrder = orderRepository.save(order)
        return savedOrder.id!!
    }

    private fun createOrderLines(
        itemResponses: List<ItemResponse>,
        orderLineRequests: List<OrderLineRequest>
    ): List<OrderItem> {
        return orderLineRequests
            .map { request: OrderLineRequest -> createOrderLine(itemResponses, request) }
            .toList()
    }

    private fun createOrderLine(itemResponses: List<ItemResponse>, orderLineRequest: OrderLineRequest): OrderItem {

        val itemResponse = findItemDtoFromOrderLineRequest(itemResponses, orderLineRequest)

        return OrderItem(
            itemResponse.id,
            itemResponse.name,
            itemResponse.size,
            itemResponse.price,
            orderLineRequest.quantity
        )
    }

    companion object {
        private fun extractItemIds(orderLineRequests: List<OrderLineRequest>): List<Long> {
            return orderLineRequests
                .map(OrderLineRequest::itemId)
                .toList()
        }

        private fun findItemDtoFromOrderLineRequest(
            itemResponses: List<ItemResponse>,
            orderLineRequest: OrderLineRequest
        ): ItemResponse {
            return itemResponses.stream()
                .filter { it: ItemResponse -> it.id == orderLineRequest.itemId }
                .findAny()
                .orElseThrow { IllegalArgumentException("주문 항목 요청에 해당하는 상품이 없습니다.") }
        }
    }
}